File: lessons/ADS/linked_list.py
from typing import Any, List, Optional, Tuple


class Node:
    def __init__(self, v: Any) -> None:
        self.value = v
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item: None) -> None:
        if self.head is None:
            self.head = item
        else:
            self.tail.next = item
        self.tail = item

    def print_all_nodes(self):
        node = self.head
        while node != None:
            print(node.value)
            node = node.next

    def find(self, val: Any, from_node: Node = None) -> Optional[Node]:
        node = self.head if from_node is None else from_node
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    def find_all(self, val: Any) -> List[Node]:
        result_list = []
        node = self.find(val)
        if node is None:
            return result_list

        result_list.append(node)
        find_more = not (node.next is None)

        while find_more:
            node = self.find(val, node.next)
            if node is None:
                break
            result_list.append(node)
            find_more = not (node.next is None)

        return result_list

    def _find_with_previous(self, val: Any) -> Tuple[Optional[Node]]:
        node = self.head
        previous_node = None
        while True:
            if node is None:
                return None, None
            if node.value == val:
                return previous_node, node
            previous_node, node = node, node.next

    def _delete(self, prev_node: Node, node: Node) -> Optional[Node]:
        if prev_node is None and node is None:
            return None
        if prev_node is None:
            self.head = node.next
        else:
            prev_node.next = node.next
        if self.tail == node:
            self.tail = prev_node
        node.next = None
        return prev_node.next if prev_node else self.head

    def delete(self, val: Any, all: bool = False) -> None:
        next_node = self.head
        while True:
            prev_node, node = self._find_with_previous(val)
            next_node = self._delete(prev_node, node)
            if not (all and next_node is not None):
                break

    def clean(self) -> None:
        while self.head:
            new_head = self.head.next
            self.head.next = None
            self.head = new_head
        self.tail = None

    def len(self) -> int:
        count = 0
        node = self.head
        while node:
            count += 1
            node = node.next
        return count

    def insert(self, after_node: Node, new_node: Node) -> None:
        new_node.next = after_node.next if after_node is not None else self.head
        if after_node is None:
            self.head = new_node
        else:
            after_node.next = new_node
        if self.tail is None or self.tail == after_node:
            self.tail = new_node

File: lessons/ads3_go/LinkedList/LinkedList.go
// 1

package main

import (
	"errors"
	"os"
	"reflect"
)

type Node struct {
	next  *Node
	value int
}

type LinkedList struct {
	head *Node
	tail *Node
}

func (l *LinkedList) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
	} else {
		l.tail.next = &item
	}

	l.tail = &item
}

// 1.5 count method. time complexity O(n); space complexity O(1)
func (l *LinkedList) Count() int {
	if l.head == nil {
		return 0
	}

	node := l.head
	for counter := 1; ; counter++ {
		node = node.next
		if node == nil {
			return counter
		}
	}
}

//1.4 find by value. time complexity O(n); space complexity O(1)
func (l *LinkedList) Find(n int) (Node, error) {
	node := l.head
	for true {
		if node == nil {
			break
		}
		if node.value == n {
			return *node, nil
		}
		node = node.next
	}
	return Node{value: -1, next: nil}, errors.New("Node not found")
}

func (l *LinkedList) findLink(n int) (*Node, error) {
	node := l.head
	for true {
		if node == nil {
			break
		}
		if node.value == n {
			return node, nil
		}
		node = node.next
	}
	return nil, errors.New("Node not found")
}

// 1.4. find all nodes by value. time complexity O(n); space complexity O(n)
func (l *LinkedList) FindAll(n int) []Node {
	nodes := []Node{}
	node := l.head
	for true {
		if node == nil {
			break
		}
		if node.value == n {
			nodes = append(nodes, *node)
		}
		node = node.next
	}
	return nodes
}

func (l *LinkedList) findAllLinks(n int) map[*Node]bool {
	nodes := map[*Node]bool{}
	node := l.head
	for true {
		if node == nil {
			break
		}
		if node.value == n {
			nodes[node] = true
		}
		node = node.next
	}
	return nodes
}

// 1.1 + 1.2 deletion of node(s) by value.
// time complexity O(n); space complexity O(1)
func (l *LinkedList) Delete(n int, all bool) {
	// look for head
	skip := false
	for true {
		if l.head == nil {
			break
		}
		if l.head.value != n {
			break
		}
		
		l.head = l.head.next
		if !all {
			skip = true
			break
		}
	}

	// go on with the rest of nodes
	parent := l.head
	if parent == nil {
		return
	}

	node := parent.next
	for true {
		if node == nil {
			break
		}
		if node.value != n || skip {
			node = node.next
			parent = parent.next
		} else {
			parent.next = node.next
			node = node.next
			if !all {
				skip = true
			}
		}
	}
	l.tail = parent
}

// 1.6 insertion after given node.
// time complexity O(n); space complexity O(1)
func (l *LinkedList) Insert(after *Node, add Node) {
	if l.tail == after {
		l.tail.next = &add
		l.tail = &add
		return
	}
	node := l.head
	for true {
		if node == nil {
			return
		}
		if node == after {
			add.next = node.next
			node.next = &add
			return
		}
		node = node.next
	}
}

// time complexity O(1); space complexity O(1)
func (l *LinkedList) InsertFirst(first Node) {
	first.next = l.head
	l.head = &first
	if l.tail == nil {
		l.tail = &first
	}
}

// 1.3 cleaning of list
// time complexity O(1); space complexity O(1)
func (l *LinkedList) Clean() {
	l.head = nil
	l.head = nil
}

